src/modes/overview/vegaSpec.js

```diff
@@ -115,10 +115,11 @@ export function createSpec(sensor, selections, dateRange, valuePatch) {
   const spec = {
     $schema: 'https://vega.github.io/schema/vega-lite/v4.json',
     data: { name: 'values' },
-    width: 'container',
-    height: 'container',
-    padding: 0,
+    width: '378',
```

Comment:
it is difficult to specify a precise number, since you don't know how wide the side panel is in every case, since it depends on the screen resolution. Especially on mobile the size is very different. You could use a similar thing `patchSpec` as in the detail view, which allow you to specify the width / height based on the container width.

Reply:
Yes, I didn't like specifying precise numbers, but I just wanted to make it work at first.   I'm not sure I like the patchSpec approach either, since it is kind of convoluted, and it requires a different call for each use.  It would be better if we could calculate it in a general way from the actual container size, however that is determined.  I'll see what I can do.

Reply:
I'm stuck.  I suspect the warning when using 'container' with autosize: 'none' is superfluous, or at least in our case.  autosize: 'none' does the right thing regarding padding, and no other mode does that, as far as I can tell.  Also, width and height set to 'container' does the right thing regarding determining the correct container size, and nothing else does that. So we need both together.  

I don't know if we care about auto sizing based on a *change* of the container size, if we can handle that another way.   If we can do that, then it would make most sense to disable the warning.  I don't see any way to do that in the vega-lite code.  This 'container' feature is new from about a year ago, so perhaps they haven't worked out all the bugs.  I can follow up with the vega-lite people to see what they think. 

Catching the log.warn and throwing it away would be one possible workaround.

Alternatively, determining the container width and height ourselves, and passing that in via something like patchSpec could work, but I haven't been able to figure out how to make that happen.  The patching has to be done differently from the DetailView case, of course, since the small multiples don't use vconcat, and it doesn't seem to work correctly if I wrap the current spec with a single vconcat so that patch could add another with size constraint.  But moreover, the spec itself isn't passed in to the Vega component the same way as for the DetailView, and I can't tell how to make it work given my limited knowledge of svelte.  Maybe you could get this working much more quickly than all my thrashing about.

Reply:
Another alternative to consider is to merge all these small multiples into one chart, using vconcat.  That is a substantial change, but I think I could handle the vega part of it.

We could save some vertical space by eliminating all but one of the domain axes, since they are now all the same, and we have the parallel selection across all of the charts.

Reply:
I added a comment to the vega-lite change that added the log warning.  See https://github.com/vega/vega-lite/pull/5441
Dominik replied, supporting my suggestion, and I would guess we should be able get this fixed at some point in vega-lite.

Meanwhile, I am wondering if we could recompute the width and height using a signals.  See https://observablehq.com/@dlaliberte/vega-top-level-width-height-padding-autosize-options  Looks like it should work at last for Vega, and I believe Vega-Lite works the same. I'll investigate.

Reply:
Using signals via the vega-embed patch appears to be working, at least partly.  I have to add separate signals for both width and height, and separate signals for both resize and load (without the load event, the initial size will be whatever the default value is).  When Switching to compare mode, the initial display of the small multiples is the wrong size, and this happens even though I am trying to use the observeResize to dispatch a window resize, in order to trigger the update of the vega chart size(s).  I'll futz with this some more to try to get it to behave, but maybe we need some more explicit resize triggers when we know there are events that need to update the charts.

By the way, the map is also not being updated based on the dispatched resizes.  Manually resizing the window does produce the desired effect (updating everything) so this is suspicious.

Reply:
Adding the following in Vega.svelte:

```
 afterUpdate(() => {
    if (!patchSpec) {
      window.dispatchEvent(new Event('resize'));
    }
  });
```

I am able to get the small multiples to behave correctly on the Map Overview page, when in Compare mode, and when resized for mobile devices, without affecting the DetailView.  However, then I learned that the overview/vegaSpec.js is also used by the top 10 and Region Details pages, which rely on different container sizes.  Fortunately, by adding the resize signals for the height as well as width, all views appear to work correctly, except for one:  When combining the DetailView with compare mode, then the DetailView chart overflows its container.  I suspect the resize events are triggering the patchSpec call to detect the wrong width at that time, before some other adjustment takes place.

Reply:
From reply to email from netlify:

Replacing the dispatch resize event with your suggested vega.view.resize().runAsync() results in a very messed up display: empty map, and it fails after trying to draw the first small multiple. 

By the way, the DetailView sizing problem also occurs even without going into "compare mode", as soon as a small multiple tooltip is shown.

I'm not seeing any performance problem with the multiple resize dispatches, so I suspect the resize events are debounced, and maybe vega is detecting that there is no size change.   However, I am seeing an afterUpdate call for each chart just due to showing a tooltip in any one of the charts.  That explains the DetailView sizing problem showing up after showing a tooltip.

So I am inclined to think that the DetailView could be fixed to avoid this problem.  I'd rather avoid doing the patchSpec call at all, if we can properly detect the container size, while also avoiding use of the arbitrary offsets to "correct" the sizing.  I'll look into that.


On Mon, Oct 26, 2020 at 8:28 AM Samuel Gratzl <notifications@github.com> wrote:
@sgratzl commented on this pull request.

In src/components/Vega.svelte:

> @@ -209,6 +237,12 @@
     }
   });
 
+  afterUpdate(() => {
+    if (!patchSpec) {
is there a difference to triggering vega.view.resize().runAsync()? cause in this in case of the small multiples N events are triggered